### PR TITLE
update current_law_policy.json for 2016 and fix a few bugs found during the process

### PR DESCRIPTION
**Pull request by MattHJensen**
This PR updates the vast majority of the params in current_law_policy.json for 2016. 
I also corrected a few bugs:

1. The child tax credit's value of 1000 was made permanent in the budget deal at the end of 2012; we had it reverting to 500 in 2018.
2. The additional child tax credit's income floor was made permanent in the budget deal at the end of last year. (We could call this a slow update rather than a bug...) 
3. `_II_brk6`, `CG_thd2`, _and, `AMT_CG_thd2` values for married filing separately included a typo in 2015 where 232425 had been mistyped as 223425. 

@martinholmer, this pr fails the requires_pufcsv tests. Could you please advise on how to proceed (note this is a repeat of the question at https://github.com/open-source-economics/Tax-Calculator/pull/550#issuecomment-175228806) 

cc @Amy-Xu, @feenberg, @GoFroggyRun 

My primary two sources are below. 
[Inflation Adjustments.pdf](https://github.com/open-source-economics/Tax-Calculator/files/106759/Inflation.Adjustments.pdf)
[irb15-44.pdf](https://github.com/open-source-economics/Tax-Calculator/files/106758/irb15-44.pdf)

### CODE REVIEW

**Comment by martinholmer** [comment]
Matt said:
> @martinholmer, this PR fails the ```requires_pufcsv``` tests. Could you please advise on how to proceed (note this is a repeat of the question at #550 (comment)).

In pull request #557, which has already been merged into the master branch, I revised what ```test_pufcsv.py``` writes out when a test fails to be (hopefully) more helpful in updating the files so that the test failures go away.  I've also added a section entitled *Updating the Test Results* at the end of the [testing procedures documentation](https://github.com/open-source-economics/Tax-Calculator/blob/master/TESTING.md).  Please let me know if all these changes answer your questions.  If not, try to explain what is unclear and I'll undertaking another round of edits.

cc @MattHJensen @feenberg @Amy-Xu @GoFroggyRun

**Comment by MattHJensen** [comment]
Thanks a lot, @martinholmer. I will try this out tomorrow morning.

**Comment by martinholmer** [comment]
Matt said:
> Thanks a lot, @martinholmer. I will try this out tomorrow morning.

@MattHJensen wait just a moment ... I'm working on a second version of the improvements.

**Comment by MattHJensen** [comment]
> @MattHJensen wait just a moment

ok. Thanks @martinholmer

**Comment by martinholmer** [comment]
Matt, I finished and merged into master branch the better ```test_pufcsv.py``` failure messaging.

So, whenever you have the time, fetch the new master branch, merge it into your 2016_clp branch, and then execute ```py.test```.  Let me know if you have any suggestions about further clarifications.

**Comment by MattHJensen** [comment]
@martinholmer, the puf testing went very smoothly; this is a big improvement to our earlier "print the diagnostic tables in the comments section" testing strategy.

**Comment by martinholmer** [comment]
Matt said:
> @martinholmer, the puf testing went very smoothly; this is a big improvement to our earlier "print the diagnostic tables in the comments section" testing strategy.

Good.  I'm pleased that you like it.  Always good to have an audit trail on how the aggregate diagnostic test results have changed with revisions to the code.

**Comment by MattHJensen** [comment]
Here is my recommendation for this PR; given that there are some bug fixes included, I recommend that @Amy-Xu give the changes a cursory review to make sure that every 2015->2016 param update looks sensible (i.e., there aren't any glaring typos) and then merge. After that, I think over the next couple of weeks @GoFroggyRun should go through all of the changes carefully, checking each one against external sources. 

>_II_brk6, CG_thd2, _and, AMT_CG_thd2 values for married filing separately included a typo in 2015 where 232425 had been mistyped as 223425.

Another topic that isn't urgent but that we should be thinking about is how to prevent bugs like this in the future, whether through process improvements or automated testing. I will open a separate issue for discussion on that.

**Comment by martinholmer** [comment]
@MattHJensen,  I'm curious about why you deleted (rather than revised) the lines just below ```# 1D data, does have 2015 values, goes up to 2018``` in the ```test_policy.py``` file.

**Comment by MattHJensen** [comment]
> @MattHJensen, I'm curious about why you deleted (rather than revised) the lines just below # 1D data, does have 2015 values, goes up to 2018 in the test_policy.py file.

After the first bugfix in the 1st PR comment, (which accounts for the fact that the $1000 child tax credit is now permanent), that child tax credit test would have duplicated the same coverage as: 

    # 1D data, doesn't have 2015 values, is not CPI inflated
    meta_kt_c_age = paramdata['_KT_c_Age']
    assert meta_kt_c_age['start_year'] == 2015
    assert meta_kt_c_age['row_label'] == ["2015"]
    assert meta_kt_c_age['value'] == [24]

**Comment by martinholmer** [comment]
@MattHJensen said:
> After the first bugfix in the 1st PR comment, (which accounts for the fact that the $1000 child tax credit is now permanent), that child tax credit test would have duplicated the same coverage as ...

OK.  That makes sense.

**Comment by MattHJensen** [comment]
@martinholmer, a downside is that there aren't any situations where the current_law_policy parameters change after 2016 anymore, but I expect there will be in the future. Our test suite won't protect us against losing the ability to handle them in the meantime. Do you have any suggestions for dealing with this lapse?

File: taxcalc/current_law_policy.json [hunk 1]
```diff
@@ -106,12 +112,14 @@
...
-		      "2014",
-		      "2015"],
+		              "2014",
+		              "2015",
+                      "2016"],
         "cpi_inflated": true,
         "col_label": "",
         "value":  [7150,
                    7250,
+                   7400,
```
Comment (Amy-Xu) [inline taxcalc/current_law_policy.json]: Just curious, where did you find the number of 7400 for 2016?

Comment (MattHJensen) [inline taxcalc/current_law_policy.json]: > 12 Alternative Minimum Tax Exemption for a Child Subject to the “Kiddie Tax.” For taxable years beginning in 2016, for a child to whom the § 1(g) “kiddie tax” applies, the exemption amount under §§ 55 and 59(j) for purposes of the alternative minimum tax under § 55 may not exceed the sum of (1) the child’s earned income for the taxable year, plus (2) $7,400.

https://www.irs.gov/pub/irs-irbs/irb15-44.pdf 

I'll add my priamry two sources as a comment in the main thread

**Comment by martinholmer** [comment]
Matt said:
> @martinholmer, a downside is that there aren't any situations where the current_law_policy parameters change after 2016 anymore, but I expect there will be in the future. Our test suite won't protect us against losing the ability to handle them in the meantime. Do you have any suggestions for dealing with this lapse?

I don't understand what you're saying.  Aren't most (or very many) tax policy parameters still being indexed for inflation under the new tax law?

**Comment by MattHJensen** [comment]
> I don't understand what you're saying. Aren't most (or very many) tax policy parameters still being indexed for inflation under the new tax law?

We used to have a few cases where the parameter was hardcoded out to a much later years than the others. We used the child tax credit for our testing of those cases (which used to be hardcoded out to 2018).  We don't have any cases like that anymore, but perhaps we don't need to be testing them.

**Comment by feenberg** [comment]
We are doing the update of the inflation adjusted dollar amounts today. 
Should be ready later this afternoon.

Can someone (Martin?)  send the 22 variable test file to me? I'd like to 
see it.

dan

**Comment by martinholmer** [comment]
Dan @feenberg said:
> We are doing the update of the inflation adjusted dollar amounts today.
> Should be ready later this afternoon.
>
> Can someone (Martin?)  send the 22 variable test file to me? I'd like to see it.

I'd be happy to send you a file, but I want to make sure I send you what you're asking for.

Do you mean you are updating the Internet-TAXSIM parameters to their actual 2015 values (rather than values indexed from some prior year), and that you want to get a set of randomly-generated filing units for 2015 in the standard Internet-TAXIM 22-variable input format?  If the answer is yes to both parts of that question, I'll be happy to send you the file named ```c15.in``` that we would use in the Internet-TAXSIM versus Tax-Calculator validation test.  I might not be able to generate this file this afternoon, but I assume getting it tomorrow will be OK.

@MattHJensen

**Comment by feenberg** [comment]
On Wed, 27 Jan 2016, Martin Holmer wrote:

> 
> Dan @feenberg said:
>
>       We are doing the update of the inflation adjusted dollar amounts today.
>       Should be ready later this afternoon.
>
>       Can someone (Martin?) send the 22 variable test file to me? I'd like to
>       see it.
> 
> I'd be happy to send you a file, but I want to make sure I send you what you're asking
> for.
> 
> Do you mean you are updating the Internet-TAXSIM parameters to their actual 2015
> values (rather than values indexed from some prior year), and that you want to get a
> set of randomly-generated filing units for 2015 in the standard Internet-TAXIM
> 22-variable input format? If the answer is yes to both parts of that question, I'll be
> happy to send you the file named c15.in that we would use in the Internet-TAXSIM
> versus Tax-Calculator validation test. I might not be able to generate this file this
> afternoon, but I assume getting it tomorrow will be OK.

Yes and yes. Is there a c14.in that has data for 2014? Any necessity for 
the input to differ from year to year? Are there any variables calculated 
by the OSPC calculator included (for my comparison)? That is, I would like 
to be able to test taxsim against OSPC, just as you test OSPC against 
taxsim.

Thanks
dan

**Comment by MattHJensen** [comment]
@Amy-Xu, does this PR look good to you?

**Comment by Amy-Xu** [comment]
Yep

On Jan 30, 2016, at 1:41 PM, Matt Jensen <notifications@github.com<mailto:notifications@github.com>> wrote:


@Amy-Xu<https://github.com/Amy-Xu>, does this PR look good to you?

**Comment by MattHJensen** [comment]
Thanks @Amy-Xu, merging.

**Pull request merged**